### PR TITLE
Generic choosers (initial implementation)

### PR DESCRIPTION
**Pull request by gasman**
First part of #8480

### CODE REVIEW

File: client/src/components/ChooserWidget/index.js [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+  getTextLabel(opts) {
+    if (!this.state) return null;
+    const result = this.state.title;
+    if (opts && opts.maxLength && result.length > opts.maxLength) {
+      return result.substring(0, opts.maxLength - 1) + '…';
+    }
+    return result;
+  }
+  focus() {
+    $('.action-choose', this.chooserElement).focus();
```
Comment (lb-) [inline client/src/components/ChooserWidget/index.js:82]: non-jquery alternative.
```suggestion
    this.chooserElement.querySelector('.action-choose').focus()
```

https://developer.mozilla.org/en-US/docs/Web/API/Element/querySelector

File: client/src/components/ChooserWidget/index.js [hunk 2]
```diff
@@ -0,0 +1,97 @@
+import $ from 'jquery';
+
+import { chooserModalOnloadHandlers } from '../../includes/chooserModal';
+
+export class Chooser {
+  modalOnloadHandlers = chooserModalOnloadHandlers;
+
+  constructor(id) {
+    this.chooserElement = $('#' + id + '-chooser');
```
Comment (lb-) [inline client/src/components/ChooserWidget/index.js:9]:
```suggestion
    this.chooserElement = document.getElementById(`${id}-chooser`);
```

File: client/src/components/ChooserWidget/index.js [hunk 3]
```diff
@@ -0,0 +1,97 @@
+import $ from 'jquery';
```
Comment (lb-) [inline client/src/components/ChooserWidget/index.js:1]:
```suggestion
```

File: client/src/components/ChooserWidget/index.js [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+import { chooserModalOnloadHandlers } from '../../includes/chooserModal';
+
+export class Chooser {
+  modalOnloadHandlers = chooserModalOnloadHandlers;
+
+  constructor(id) {
+    this.chooserElement = $('#' + id + '-chooser');
+    this.titleElement = this.chooserElement.find('.title');
+    this.input = $('#' + id);
```
Comment (lb-) [inline client/src/components/ChooserWidget/index.js:11]:
```suggestion
    this.input = document.getElementById(id);
```

File: client/src/components/ChooserWidget/index.js [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+    this.editLink = this.chooserElement.find('.edit-link');
```
Comment (lb-) [inline client/src/components/ChooserWidget/index.js:12]:
```suggestion
    this.editLink = this.chooserElement.querySelector('.edit-link');
```

File: client/src/components/ChooserWidget/index.js [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    this.chooserBaseUrl = this.chooserElement.data('chooserUrl');
```
Comment (lb-) [inline client/src/components/ChooserWidget/index.js:13]:
```suggestion
    this.chooserBaseUrl = this.chooserElement.dataset.chooserUrl;
```

File: client/src/components/ChooserWidget/index.js [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+      this.renderState(newState);
+    } else {
+      this.renderEmptyState();
+    }
+  }
+  clear() {
+    this.setState(null);
+  }
+  renderEmptyState() {
+    this.input.val('');
```
Comment (lb-) [inline client/src/components/ChooserWidget/index.js:63]:
```suggestion
    this.input.setAttribute('value', '');
```

File: client/src/components/ChooserWidget/index.js [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+    this.chooserElement.addClass('blank');
+  }
+  renderState(newState) {
+    this.input.val(newState.id);
```
Comment (lb-) [inline client/src/components/ChooserWidget/index.js:67]:
```suggestion
    this.input.setAttribute('value', newState.id);
```

File: client/src/components/ChooserWidget/index.js [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+    this.titleElement.text(newState.title);
```
Comment (lb-) [inline client/src/components/ChooserWidget/index.js:68]:
```suggestion
    this.titleElement.innerText = newState.title;
```

File: client/src/components/ChooserWidget/index.js (hunk 3 above)
Comment (lb-) [inline client/src/components/ChooserWidget/index.js:1]: I know there is not a clear guideline yet but not sure this should be in the `components` folder.

Again, this should probably be in the `includes` folder for now as this is not a React component.

Also, just curious if there is a possibility to make this typescript (I am happy to help with this).

**Review by lb- — COMMENTED** [review]
Just added a bunch of suggested changes to move away from jQuery, let me know if you would rather me add a commit to your branch.

I know this is a bit annoying but hoping we can move away from reliance on jQuery for new code and ideally use TypeScript if possible.

Note: I have not run the recommended changes locally but hopefully should be pretty close :)

**Comment by gasman** [comment]
@lb- Thanks for the feedback! I'm all in favour of eliminating jQuery here - I think I looked at that briefly in some of the earlier refactoring, but in the case of the modals it couldn't be done easily without changing APIs because modal-workflow.js returns a jQuery object for `modal.body`. That's not an issue for the chooser widget though... happy to go through and apply your suggestions.

File: client/src/components/ChooserWidget/index.js (hunk 3 above)
Comment (gasman) [reply client/src/components/ChooserWidget/index.js:1]: FWIW `components/StreamField` sets a precedent of breaking that rule (which is not to say it's a good thing to break it again)... I chose `components` here as it seemed more conceptually component-y than the other code (being an identifiable element on the page with JS behaviour). Happy to go with whatever conventions we establish though.

**Comment by gasman** [comment]
jQuery now eliminated as suggested.

File: docs/reference/viewsets.md [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+-------
+
+.. autoclass:: wagtail.admin.viewsets.base.ViewSet
+
+   .. automethod:: on_register
+   .. automethod:: get_urlpatterns
+   .. automethod:: get_url_name
+
+ModelViewSet
+------------
```
Comment (lb-) [inline docs/reference/viewsets.md:17]: Would it be possible to split out each code autoclass into its own eval-rst block and then put the headers in markdown?

**Review by lb- — APPROVED** [review]
Looks good @gasman  I will merge this in.

I will assume that any unit tests will be done as part of the larger work.

**Pull request merged**